### ISSUE TITLE
Product with Reserved stock shown as "in stock"

### ISSUE THREAD

**Issue by glagonikas** [woocommerce/woocommerce#29519]
**Prerequisites (mark completed items with an [x]):**
- [X] I have have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**
A product with limited stock, where all remaining units are reserved in unpaid orders, is displayed as "In stock" in the product page, allowing customers to add it to their basket. Once they add it and move to the basket or checkout, they see a message saying "Sorry, we do not have enough "Product" in stock to fulfill your order (0 available). We apologise for any inconvenience caused."

**Expected behavior**
The product should be displayed as "Out of stock" while all units are reserved.
There is no reason to allow users to add products to their cart if they cannot buy them. Makes for bad UX.
Some sites might turn off hold stock (or reduce it to a few minutes) but if they do abandoned carts (AutomateWoo) they need to retain the stock in the carts for days to allow for customers to return to their carts.

**Actual behavior**
The product is displayed as "In Stock" and the customer can add it to their basket.
When they move to their basket or checkout, they see an error message

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**
Steps to reproduce the bug:
1. Create a new product
2. Set stock to 1
3. Go to that product as a customer, add to your cart, go to checkout, and create an unpaid order. Close browser
4. Starting a new session, go to the product page. See how the product is showing as "in stock"
5. Click to add the product to your basket
6. Go to your basket

**Screenshots**
Product page when all units are reserved
![screenshot-2021 03 29-11_05_26](https://user-images.githubusercontent.com/10516403/112823951-d4c82700-9081-11eb-83fd-bf7ca684b437.png)

Cart page when all units are reserved
![screenshot-2021 03 29-11_05_50](https://user-images.githubusercontent.com/10516403/112823924-cbd75580-9081-11eb-90d5-d8b98a229124.png)

**Isolating the problem (mark completed items with an [x]):**
- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by juliaamosova** [comment]
Hi @glagonikas,

Thank you for submitting the issue. However, I can’t reproduce it using the steps you provided. Everything is working as expected on my end. I tested using WooCommerce 5.1. 

>Go to that product as a customer, add to your cart, go to checkout, and create an unpaid order. Close browser

Can you please specify what kind of unpaid order should be created in order to encounter the issue? 

I created an unpaid order using Direct Bank Transfer payment where the order is placed but it is unpaid. As a result, if I visit the product page, I am getting a proper `Out of stock` message as expected:

<img width="1261" alt="bacs" src="https://user-images.githubusercontent.com/19143190/113487687-805ce700-9487-11eb-8af9-5f0c66ae30b5.png">

<img width="961" alt="bacs2" src="https://user-images.githubusercontent.com/19143190/113487692-84890480-9487-11eb-982d-bff84f74855f.png">

Please provide us with more details about the issue which may help us to evaluate it further.

**Comment by glagonikas** [comment]
Hi @juliaamosova I was testing using PayPal, where the order is created and set to "pending" instead of "on hold" and is not paid.

In our case, customers are sent to PayPal but for whatever reason, (forgot their PayPal password/not enough money/changed mind) they drop without completing their order. Those orders do not contain any notes.

**Comment by AashikP** [comment]
Hello @glagonikas 

Thank you for getting back to us with clarification on that, we really appreciate your help. I can reproduce it on my end. 

I tried to replicate this behavior on my test site, where the customer attempts to pay via PayPal and then does not proceed with the payment (I closed the PayPal/Checkout page when I was redirected).

This resulted in an order with no order notes:

![https://d.pr/i/4iwA0o+](https://d.pr/i/4iwA0o+)

However, no stock was reduced on the product itself:

![https://d.pr/i/67LggB+](https://d.pr/i/67LggB+)

I could also add the product to the cart when accessing the same product via an incognito window:

![https://d.pr/i/mLylI0+](https://d.pr/i/mLylI0+)

However, the cart page shows the error as you mentioned:

![https://d.pr/i/trKyCS+](https://d.pr/i/trKyCS+)

---

As you clarified, the product should indeed be displayed as out of stock until the `hold stock` time is reached, and the original order is marked as canceled. 

Priority of this issue will be assigned shortly after we'd go through a round of prioritization. Please note it may take a couple of days.

**Comment by grig-23** [comment]
Interesting! I did a few tests, investigating a slightly related topic, and here are my findings:
 
I had 4 remaining items in stock. I added 3 to the cart and placed an order with PayPal Standard. The order status  was set to “pending payment”. Stock isn’t reduced. But, I still see the products in cart! Therefore, I can’t add 3 or 4 again - it tells me `You cannot add that amount to the cart — we have 4 in stock and you already have 3 in your cart.`
 
Then, in incognito mode (as a different user) I tried adding 4 to the cart - I was able to do that, but at the cart page I see `Sorry, we do not have enough "Simple product" in stock to fulfill your order (1 available). We apologize for any inconvenience caused.`

<img width="966" alt="Screenshot 2022-07-29 at 17 39 13" src="https://user-images.githubusercontent.com/13084901/181796082-a18e2e6a-cee5-446d-bc30-f98cefacd2b9.png">

I'm pretty sure that this is how the Hold Stock setting works - it doesn't reduce stock, but it prevents other users from adding the same items to their cart, because these items are in someone else's cart.

Tested with WooCommerce 6.6.0.

**Comment by glagonikas** [comment]
@grig-23 for clarity, I think your findings are slightly different to this issue and are expected behaviour (sort of).

In your tests, you are able to add 4 items to your second cart because the item is still in stock since there is at least 1 item available. This is not great UX (no errors are great UX), but it's understandable. Ideally, WC should allow you to pick a quantity up to the available stock amount, so in your case, it should only allow you to pick one.

This issue refers to customers who can add to their carts, even though there are no units available. so if all units are reserved, customers should see the "out of stock" message instead of having to go to the cart, and then be told there are no more available.

**Comment by pmcpinto** [comment]
It looks like this is still relevant. I've marked it as ready for engineering triage.

Current evidence: Current trunk still shows stock and gates add-to-cart on `is_in_stock()` / `wc_get_stock_html()` at the product page, while cart validation separately subtracts held stock with `wc_get_held_stock_quantity()`. That leaves a clear gap where fully reserved stock can still appear purchasable on the PDP and only fail later in cart or checkout.

**Comment by verofasulo** [comment]
That's such a good discussion and a tricky one. 

I'll try to summarize it to see if I understood the problem correctly.

### **The problem**
When stock is temporarily held for an in-progress or unpaid order, the storefront can still advertise the item as available — so a second shopper is told "yes," then "no" at checkout.

### **Why WooCommerce hits this gap**

WooCommerce stores availability as a single stored quantity plus a status flag (in stock / out of stock). Reservations — both the short checkout hold and the "hold stock for unpaid orders" setting — are tracked separately and are never folded back into that stored status.

- The product page reads the stored number → sees 1 → shows "In stock."
- The purchase check reads stored quantity minus active reservations → 1 − 1 = 0 → rejects the sale.

Those two answers come from two different calculations, and they disagree.

### **How Shopify handles it**

Shopify doesn't keep one number and a flag. It models several states at once:

> On hand = Committed + Unavailable + Available

The storefront always sells against **Available** ("inventory you can sell… not committed to any orders or set aside as unavailable"). The moment a unit is committed to a placed order — or reserved during an active checkout — it leaves Available, so the product shows sold out automatically. There is no separate status flag to fall out of sync: **the number the shopper sees is the number that gates the purchase**. Shopify also reserves stock for a limited window during checkout to stop concurrent shoppers grabbing the same last unit.

### **How Wix handles it**

Wix has no cart or checkout hold. Inventory drops at one configurable moment:

- "When the order is placed" — decrement as soon as checkout completes, or
- "Only after the order is paid."

Because nothing is ever held mid-flow, Wix never has a "reserved-but-shown-as-available" gap. The trade-off: with "after paid," high traffic plus slow payment providers can let customers buy out-of-stock items.

----

WooCommerce sits in the middle: it reserves like Shopify but displays like Wix's raw-quantity model. 

The Shopify approach seems to be clearer to me, but that would require a big architectural shift. For now, we could do the following:

### Stage 1 — Single product page

This is where the worst version happens: a shopper reads the page, adds the item, and only then gets turned away. So the smallest useful fix is to make that one page subtract the set-aside units before it says "In stock."

### Stage 1 — Product form

Right now the set-aside units are invisible to the shop owner, so if a product suddenly says "sold out" when they know they have some, they'll think something's broken. We can show them the breakdown:

- **On the product's edit screen**: a small line under the Quantity box, shown only when some units are on hold — for example, "16 in stock — 5 held by unpaid orders, 11 available."

<img width="1778" height="1260" alt="Image" src="https://github.com/user-attachments/assets/c4a8edef-5a8a-4375-a132-7feed0da3267" />

- All Products list → Stock column: a compact held/available indicator. I haven't put together any designs yet; let's see if it's possible in the product editor first. 

What do you think?
